EIPS/eip-3224.md

```diff
@@ -0,0 +1,465 @@
+author: Richard Moore (@ricmoo), Nick Johnson (@arachnid)
+discussions-to: https://github.com/ethereum/EIPs/issues/3225
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-01-23
+requires: 191
+---
+
+## Simple Summary
+
+A technique for contract authors to enable wallets to provide
+a human-readable description of what a given contract
+interaction (via a transaction or signed message) will perform.
+
+
+## Abstract
+
+Human-readable descriptions for machine executable operations,
+described in higher level machine readable data, so that wallets
+can provide meaningful feedback to the user describing the
+action the user is about to perform.
+
+
+## Motivation
+
+When using an Ethereum Wallet (e.g. MetaMask, Clef, Hardware
+Wallets) users must accept and authorize signing messages or
+sending transactions.
+
+Due to the complexity of Ethereum transactions, wallets are very
+limitd in their ability to provide insight into the contents of
+transactions user are approving; outside special-cased support
+for common transactions such as ERC20 transfers, this often amounts
+to asking the user to sign an opaque blob of binary data.
+
+This EIP presents a method for dapp developers to enable a more
+comfortable user experience by providing wallets with a means
+to generate a better description about what the contract claims
+will happen.
+
+It does not address malicious contracts which wish to lie, it
+only addresses honest contracts that want to make their user's
+life better. We believe that this is a reasonable security model,
+as transaction descriptions can be audited at the same time as
+contract code, allowing auditors and code reviewers to check that
+transaction descriptions are accurate as part of their review.
+
+
+## Specification
+
+The **description string** and **described data** are generated
+simultaneously by evaluating **describer** bytecode, passed
+**describedInput** as the calldata with the encoded method:
+
+```solidity
+function eipXXXDescribe(bytes inputs) view returns (string description, bytes data);
+```
+
+The method MUST be executable in a static context, i.e. any
+side effects (logX, sstore, etc., (including through indirect
+calls) will be ignored.
+
+During evaluation, the `ADDRESS` (i.e. `to`), `CALLER`
+(i.e. `from`), `VALUE`, and `GASPRICE` must be the same as the
+values for the transaction being described, so that the
+code generating the description can rely on them. For signing
+**described messages**, `VALUE` should always be 0.
+
+When executing the bytecode, best efforts should be made to
+ensure `BLOCKHASH`, `NUMBER`, `TIMESTAMP` and `DIFFICULTY`
+match the `"latest"` block. The `COINBASE` should be the zero
+address.
+
+If evaluating the function results in a revert, and the return
+data is of a length that is congruent to `4 mod 32`, and the
+first 4 bytes represent the selector `Error(string)`, the
+reason string should be displayed to the user. If evaluation
```

Comment:
Displayed where?  Also, in general `should` is a red flag in a standard as anything that is a `should` or `may` cannot be enforced for compatibility.  This is especially true in this case where I *think* you are trying to tell user interfaces what they should do, which is more appropriate for an opinion piece like a blog post, tutorial, best practices document, patterns book, etc.

Reply:
I guess the reason for the *should* is that not all UI's are interactive or have a display.

Reply:
This paragraph feels out of scope for an ERC.  UIs can do whatever they want once they have completed the interaction with the contract.  I do agree that what you have described is a Good Idea, but EIPs are for standardization, not Good Ideas.  😄

Reply:
I've removed this and simple mentioned that the method may revert and that if so, the signing must be aborted.